File: 0006-cycle-linked-list/0006-cycle-linked-list.java
// HackerRank provided the other necessary codes to run

class HasCycle {

  static boolean hasCycle(SinglyLinkedListNode head) {
    Map<SinglyLinkedListNode, Boolean> visitedNodes = new HashMap<>();
    SinglyLinkedListNode node = head;
    while (node != null) {
      if (visitedNodes.get(node) != null) {
        return true;
      }
      visitedNodes.put(node, true);
      node = node.next;
    }
    return false;
  }
}

File: 0006-cycle-linked-list/0006-cycle-linked-list.ts
class SinglyLinkedListNode {
    data: any;
    next: SinglyLinkedListNode | null;

    constructor(data: any) {
        this.data = data;
        this.next = null;
    }
}

class HasCycle {
    static hasCycle(head: SinglyLinkedListNode | null): boolean {
        const visitedNodes: Map<SinglyLinkedListNode, boolean> = new Map();
        let node: SinglyLinkedListNode | null = head;

        while (node !== null) {
            if (visitedNodes.has(node)) {
                return true;
            }
            visitedNodes.set(node, true);
            if (node.next !== null) {
                node = node.next;
            } else {
                break;
            }
        }
        return false;
    }
}
